### PR TITLE
Query bank balances

### PR DESCRIPTION
**Pull request by samsiegart**
refs https://github.com/Agoric/wallet-app/issues/49

We still need to update dapp-inter to do the same, then I can test e2e with amounts in offer proposals rather than pursepetname+value

### CODE REVIEW

File: src/util/WalletBackendAdapter.ts
Comment (samsiegart) [inline src/util/WalletBackendAdapter.ts:233]: If an asset exists in the vbank we want to show it as a purse with 0 balance so they can click the eventual "deposit" button for ibc transfer.

File: src/service/Offers.ts
Comment (dckc) [inline src/service/Offers.ts:53]: we're cool with open-coding `AmountMath.make()`? maybe the notion of `brand` used here doesn't quite line up?

File: src/util/queryBankBalances.ts
Comment (dckc) [inline src/util/queryBankBalances.ts:11]: ambient authority. hm.

> No modules should export powerful objects (for example, objects that close over fs or process.env). -- https://github.com/Agoric/agoric-sdk/wiki/OCap-Discipline

I suppose we have agreed on an exception to that for UI code? I'd like to see that in the README or the like.

File: src/util/queryBankBalances.ts
Comment (dckc) [inline src/util/queryBankBalances.ts:16]: **scaling consideration**: How often do we do this? I gather from @arirubinstein that we're supposed to make no more than 2 RPC queries per second.

I gather @michaelfig plans to provide rate limiting in `@agoric/casting` (IOU an issue #) Do we have an issue about migrating this code to use it? Cite it from a TODO?

File: src/util/WalletBackendAdapter.ts
Comment (dckc) [inline src/util/WalletBackendAdapter.ts:306]: This documentation is good.

Ideally we'd import some `invitationIssuerAssetKind` constant from `@agoric/zoe/something`.

I leave it to @turadg to judge the cost-effectiveness of that.

**Review by dckc — COMMENTED** [review]
Looks plausible.

I'd like to see test coverage for this feature, as usual. Skipping it for now is OK with me if it's OK with @turadg .

File: src/util/queryBankBalances.ts
Comment (samsiegart) [reply src/util/queryBankBalances.ts:16]: Yea my hope is this could unblock us for now and we can swap in the improved casting stuff later, just added a comment here.

File: src/util/queryBankBalances.ts
Comment (samsiegart) [reply src/util/queryBankBalances.ts:11]: Noted, I don't know if I have a clear way to explain the philosophy as it pertains to UI code at the moment though.

File: src/service/Offers.ts
Comment (samsiegart) [reply src/service/Offers.ts]: Thanks, you caught a bug, I wasn't converting value to BigInt. Just changed to use AmountMath and confirmed it works.

File: src/service/Offers.ts
Comment (turadg) [inline src/service/Offers.ts:47]: this is a little hard to follow. consider filtering and naming the value transformer.

```
const convertProposal = ({ pursePetname, value, amount: serializedAmount }) =>
  serializedAmount
              ? await E(boardIdMarshaller).unserialize(serializedAmount)
              : AmountMath.make(
                  pursePetnameToBrand.get(pursePetname),
                  BigInt(value);

Object.entries(paymentProposals).
filter(([_kw, { pursePetname, amount}]) => amount && pursePetnameToBrand.get(pursePetname))
.map(([kw, proposal]) => convertProposal);
```

Then it's a small step to using `objectMap` and `deeplyFulfilled`;

File: src/service/Offers.ts
Comment (turadg) [inline src/service/Offers.ts:55]: thanks for the documenting. consider TODO.

File: src/service/Offers.ts
Comment (turadg) [inline src/service/Offers.ts:74]: i had to remember there's no promise pipelining in this app so the two awaits are fine.

it wasn't obvious to me why you wanted to wrap these in a function. It looks like you want `convertProposals` to resolve in parallel to resolving the `instance`. Good idea. But I don't see why you want `instanceBoardId` to be available after this Promise.all.

consider,
```
    const [instance, give, want] = await Promise.all([
      E(boardIdMarshaller).unserialize(instanceHandle),
      convertProposals(giveTemplate),
      convertProposals(wantTemplate),
    ]);
```
and then just above the first use of `instanceBoardId`,
```
    const {
      slots: [instanceBoardId],
    } = await E(boardIdMarshaller).serialize(instance);
```

File: src/util/WalletBackendAdapter.ts
Comment (turadg) [inline src/util/WalletBackendAdapter.ts:218]: consider "Infer purses from Cosmos bank module balances" since, as you say, purses can't be read from Cosmos.

File: src/util/WalletBackendAdapter.ts
Comment (turadg) [reply src/util/WalletBackendAdapter.ts:233]: Very helpful note. Please include in the code.

File: src/util/WalletBackendAdapter.ts
Comment (turadg) [inline src/util/WalletBackendAdapter.ts:251]: why the const?
```suggestion
      bank = await queryBankBalances(keplrConnection.address, rpc);
```

File: src/util/WalletBackendAdapter.ts
Comment (turadg) [inline src/util/WalletBackendAdapter.ts:270]: when would the iterator return an object without a `value`?

File: src/util/WalletBackendAdapter.ts
Comment (turadg) [inline src/util/WalletBackendAdapter.ts:271]: This took me some time to read through without type info. I think it's making a map with the keys/values inverted. Consider https://lodash.com/docs/4.17.15#invert to make that clear, or a code comment.

File: src/util/WalletBackendAdapter.ts
Comment (turadg) [inline src/util/WalletBackendAdapter.ts:129]: please include the types. if they can't be imported just write out what this code is assuming.

Comment (turadg) [reply src/util/WalletBackendAdapter.ts:129]: also this is getting very long for a positional arguments list. very much worth turning into a named options argument, though not necessarily this PR

File: src/util/WalletBackendAdapter.ts
Comment (turadg) [reply src/util/WalletBackendAdapter.ts:306]: I don't know of any runtime features of an invitation issuer that distinguishes it from any other Set kind issuer.

You could make this "set"-ness check more apparent using `@agoric/patterns` (soon to be in Endo): `matches(purse.balances.value, SetValueShape)`.

File: src/util/WalletBackendAdapter.ts
Comment (turadg) [inline src/util/WalletBackendAdapter.ts:312]: consider distinguishing these conditions:
```
if (!Array.isArray(purse.balance.value)) {
  console.debug('skipping non-set amount', purse.balance.value);
}
if (!agoricBrands.has(purse.brand)) {
  console.debug('skipping unknown brand', purse.brand);
```
Likely to pay off debugging someday. And in the meantime it's code documentation.

File: src/util/queryBankBalances.ts
Comment (turadg) [inline src/util/queryBankBalances.ts:7]: Do we have a ticket to track when it's possible? I think we should. Consider noting it here with an `UNTIL`.

File: src/util/queryBankBalances.ts
Comment (turadg) [reply src/util/queryBankBalances.ts:13]: I think we have agreed agreed on an exception to that for UI code and I agree it would be valuable to have in the README of this repo to refer to. Not a blocker imo

File: yarn.lock
Comment (turadg) [inline yarn.lock:12167]: curious.

File: src/service/Offers.ts
Comment (samsiegart) [reply src/service/Offers.ts]: Done.

I'm not sure about promise pipelining in the rest of the app, but we await here so that the card doesn't show up before it has the data needed to render.

File: src/service/Offers.ts
Comment (samsiegart) [reply src/service/Offers.ts]: Done

File: src/service/Offers.ts
Comment (samsiegart) [reply src/service/Offers.ts:47]: I gave this a shot but I got a little confused with how to use those utility functions and the code wasn't looking much simpler, I'd prefer to leave this as is.

File: src/util/WalletBackendAdapter.ts
Comment (samsiegart) [reply src/util/WalletBackendAdapter.ts]: Added types here and a bit below.

File: src/util/WalletBackendAdapter.ts
Comment (samsiegart) [reply src/util/WalletBackendAdapter.ts]: Done

File: src/util/WalletBackendAdapter.ts
Comment (samsiegart) [reply src/util/WalletBackendAdapter.ts:244]: Done

File: src/util/WalletBackendAdapter.ts
Comment (samsiegart) [reply src/util/WalletBackendAdapter.ts]: Oops, done

File: src/util/WalletBackendAdapter.ts
Comment (samsiegart) [reply src/util/WalletBackendAdapter.ts]: I guess I was afraid it might have an initial value of null or something, but that's more a concern for notifiers, and it would throw anyway in that case. Just removed the conditional

File: src/util/WalletBackendAdapter.ts
Comment (samsiegart) [reply src/util/WalletBackendAdapter.ts]: Added types and left a comment

File: src/util/WalletBackendAdapter.ts
Comment (samsiegart) [reply src/util/WalletBackendAdapter.ts]: Looks like `@agoric/patterns` doesn't exist yet either, but that would be cool.

File: src/util/WalletBackendAdapter.ts
Comment (samsiegart) [reply src/util/WalletBackendAdapter.ts]: Done

File: src/util/WalletBackendAdapter.ts
Comment (dckc) [reply src/util/WalletBackendAdapter.ts:129]: I was going to say something about the `unknown` type, but if it typechecks (without casts), that means the code doesn't care what the type is, right? i.e. the code is assuming nothing.

File: src/util/queryBankBalances.ts
Comment (samsiegart) [reply src/util/queryBankBalances.ts]: Not sure about the ticket yet, changed to `UNTIL`

File: src/util/queryBankBalances.ts
Comment (dckc) [reply src/util/queryBankBalances.ts:13]: agreed, not a blocker.

meanwhile, I realize pointing to the wiki doesn't automatically make a link the other way, where citing an issue does: https://github.com/Agoric/agoric-sdk/issues/2160

File: src/util/WalletBackendAdapter.ts
Comment (turadg) [reply src/util/WalletBackendAdapter.ts]: Oops I meant '@agoric/store'. But yeah let's wait for it to be in Endo

**Review by turadg — APPROVED** [review]
🚀

File: src/util/queryBankBalances.ts
Comment (dckc) [reply src/util/queryBankBalances.ts:18]: filed:
 - https://github.com/Agoric/agoric-sdk/issues/6870

**Pull request merged**
